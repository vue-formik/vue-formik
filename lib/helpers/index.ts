/* eslint-disable @typescript-eslint/no-explicit-any */
const getNestedValue = <T extends object>(obj: T, path: string) => {
  if (!obj || typeof obj !== "object") return undefined;

  if (path in obj) {
    return obj[path as keyof T];
  }

  const keys = path.split('.'); // Split the path into keys

  if (keys.length === 1) {
    if (/^.*\[\d+]$/.test(keys[0])) {
      const [key, index] = keys[0].split("[");
      const i = parseInt(index.replace("]", ""), 10);
      const element = obj[key as keyof T];
      if (Array.isArray(element)) {
        return element[i];
      } else {
        return undefined;
      }
    }
    return obj[keys[0] as keyof T];
  }

  return getNestedValue(obj[keys[0] as keyof T] as T, keys.slice(1).join('.')); // Recursively get the nested value
};

const updateNestedProperty = <T extends Record<string, any>>(
  obj: T,
  path: string,
  value: any,
): void => {
  if (!obj || typeof obj !== "object") {
    obj = {} as T;
    obj[path as keyof T] = value;
    return;
  }

  if (path in obj) {
    // update object here
    obj[path as keyof T] = value;
    return;
  }

  const keys = path.split('.');

  if (keys.length === 1) {
    if (/^.*\[\d+]$/.test(keys[0])) {
      const [key, index] = keys[0].split("[");
      if (!Array.isArray(obj[key as keyof T])) {
        obj[key as keyof T] = [];
      }
      const i = parseInt(index.replace("]", ""), 10);

      obj[key as keyof T][i] = value;
      return;
    }
    obj[keys[0] as keyof T] = value;
    return;
  }
  // Split the path into keys
  return updateNestedProperty(getNestedValue(obj, keys[0]), keys.slice(1).join('.'), value); // Recursively get the nested value
};

function clearReactiveObject<T extends object>(obj: T): void {
  if (typeof obj !== "object" || obj === null) {
    return; // Skip non-object or null values
  }

  // Skip unsupported types like Map, Set, WeakMap, WeakSet
  if (
    obj instanceof Map ||
    obj instanceof WeakMap ||
    obj instanceof Set ||
    obj instanceof WeakSet
  ) {
    return;
  }

  // Check for Vue's reactive types and unwrap them if needed
  if ("__v_isRef" in obj || "__v_isReactive" in obj) {
    return;
  }

  for (const key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      const value = obj[key as keyof T];
      if (value && typeof value === "object" && !Array.isArray(value)) {
        clearReactiveObject(value); // Recursively clear nested objects
      }
      delete obj[key as keyof T]; // Delete the property
    }
  }
}

export { updateNestedProperty, getNestedValue, clearReactiveObject };
