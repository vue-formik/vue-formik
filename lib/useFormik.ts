import { computed, reactive, toRaw, watch } from 'vue'
import { ObjectSchema } from 'yup'
import type { FormikHelpers, IUseFormik, ValidationRule } from '@/types'
import { getNestedValue, updateNestedProperty } from '@/helpers'

const useFormik = <T extends object>({
  initialValues,
  validationSchema,
  onSubmit,
}: IUseFormik<T>) => {
  const validate = (values: T): Partial<Record<keyof T, string>> => {
    const validationErrors: Partial<Record<keyof T, string>> = {};

    if (validationSchema instanceof ObjectSchema) {
      try {
        const vSchema = validationSchema as ObjectSchema<T>;
        vSchema.validateSync(values, { abortEarly: false });
      } catch (e) {
        const err = e as never as { inner: { path: string; message: string }[] };
        if (typeof err === "object" && Array.isArray(err?.inner)) {
          err.inner.forEach((error) => {
            validationErrors[error.path as keyof T] = error.message;
          });
        }
      }
    } else {
      for (const key in validationSchema) {
        const value = values[key as keyof T];
        const rule = validationSchema[key as keyof T] as ValidationRule<T[keyof T]>;

        if (typeof rule === 'function') {
          const error = rule(value);
          if (error) {
            validationErrors[key as keyof T] = error;
          }
        }
      }
    }

    return validationErrors;
  };

  const values = reactive(initialValues);
  const errors = reactive(validate(initialValues));
  const touched = reactive({} as Partial<Record<keyof T, boolean | Record<string, boolean>>>);

  const setValues = (newValues: T) => {
    Object.assign(values, newValues); // Safely copy values into reactive object
    Object.assign(errors, validate(newValues)); // Validate updated values
  };

  const setErrors = (newErrors: T) => {
    Object.assign(errors, newErrors);
  };

  const reset = () => {
    Object.assign(values, initialValues);
    Object.assign(errors, validate(initialValues));
  };

  const setFieldValue = (field: string, value: unknown) => {
    updateNestedProperty(values as Record<string, unknown>, field, value);
  };

  const setFieldTouched = (field: string, touchedValue: boolean) => {
    updateNestedProperty(touched as Record<string, unknown>, field, touchedValue);
  }

  const handleSubmit = (e: Event) => {
    e.preventDefault();
    onSubmit(
      toRaw(values) as T,
      {
        setErrors,
        reset,
        setValues,
      } as FormikHelpers<T>
    );
  };

  const handleBlur = (e: FocusEvent) => {
    const fieldName = (e.target as HTMLInputElement).name as keyof T;
    setFieldTouched(fieldName.toString(), true);
  }
  const handleChange = (e: Event) => {
    const target = e.target as HTMLInputElement;
    const fieldName = target.name as keyof T;
    const value = target.type === 'checkbox' ? target.checked : target.value;

    setFieldValue(fieldName.toString(), value);
  }

  const isValid = computed(() => {
    return Object.keys(errors).length === 0;
  })

  const isDirty = computed(() => {
    return JSON.stringify(values) !== JSON.stringify(initialValues);
  })

  watch(
    () => values,
    () => {
      const validationErrors = validate(toRaw(values) as T);
      //@ts-expect-error Object.keys(errors) is not iterable
      Object.keys(errors).forEach((key) => delete errors[key]); // Clear existing errors
      Object.assign(errors, validationErrors); // Assign new errors
    },
    { deep: true }
  );


  const hasFieldError = (field: string) => {
    const errorValue = getNestedValue(errors, field);
    const touchedValue = getNestedValue(touched, field);
    return (errorValue !== undefined && !!touchedValue);
  }

  const getFieldError = (field: string) => {
    if (hasFieldError(field)) {
      return errors[field as keyof typeof errors];
    } else {
      return '';
    }
  }

  return {
    values,
    errors,
    touched,
    setValues,
    reset,
    setFieldValue,
    setFieldTouched,
    handleBlur,
    handleChange,
    handleSubmit,
    hasFieldError,
    getFieldError,
    isValid,
    isDirty
  };
};

export default useFormik;
