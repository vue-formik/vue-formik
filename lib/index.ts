import useFormik from "@/composables/useFormik";
import FormTextArea from "@/components/FormTextArea.vue";
import FormTextField from "@/components/FormTextField.vue";

export { FormTextArea, FormTextField, useFormik };
