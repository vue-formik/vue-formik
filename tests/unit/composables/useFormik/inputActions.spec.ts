import { describe, expect, it } from "vitest";
import { useFormik } from "../../../../lib";

describe("useFormik input actions", async () => {
  describe("handleChange", () => {
    /**
     *   const handleFieldChange = (e: Event) => {
     *     const target = e.target as HTMLInputElement;
     *     const fieldName = target.name as keyof T;
     *     const value = target.type === "checkbox" ? target.checked : target.value;
     *
     *     setFieldValue(fieldName, value as T[keyof T]);
     *   };
     */
    const cases = [
      // {
      //   target: {
      //     name: "contacts",
      //     type: "text",
      //     value: "9876543210",
      //   },
      //   initialValues: { contacts: ""},
      //   expectedKey: "contacts",
      // },
      // {
      //   target: {
      //     name: "contacts[0]",
      //     type: "text",
      //     value: "9876543210",
      //   },
      //   initialValues: { contacts: [""]},
      //   expectedKey: "contacts[0]",
      // },
      {
        target: {
          name: "contacts[0].number",
          type: "text",
          value: "9876543210",
        },
        initialValues: { contacts: [{ code: "", number: "" }] },
        expectedKey: "contacts[0].number",
      },
    ];

    it.each(cases)("should set field value", async ({target, initialValues}) => {
      const { handleChange, values } = useFormik({ initialValues });

      // Act
      handleChange({ target });

      // Assert
      expect(values).toMatchSnapshot();
    })
    it.each(cases)("should set the field as touched", ({target, initialValues, expectedKey}) => {
      const { handleChange, touched } = useFormik({ initialValues });

      // Act
      handleChange({ target });

      // Assert
      expect(touched).toMatchSnapshot();
    })
  })
})
