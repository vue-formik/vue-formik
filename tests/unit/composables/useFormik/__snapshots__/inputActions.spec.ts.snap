// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFormik input actions > handleChange > should set field value 1`] = `
{
  "contacts": "9876543210",
}
`;

exports[`useFormik input actions > handleChange > should set field value 2`] = `
{
  "contacts": [
    "9876543210",
  ],
}
`;

exports[`useFormik input actions > handleChange > should set field value 3`] = `
{
  "contacts": [
    {
      "code": "",
      "number": "9876543210",
    },
  ],
}
`;

exports[`useFormik input actions > handleChange > should set the field as touched 1`] = `{}`;
